File: Dynamic Programming/Python/Best Sightseeing Pair.py
# Problem link: https://leetcode.com/problems/best-sightseeing-pair/description
# Time: O(n)
# Space: O(1)
class Solution:
    def maxScoreSightseeingPair(self, values: List[int]) -> int:
        currVal, result = values[0] - 1, 0

        for index in range(1, len(values)):
            result = max(result, currVal + values[index])
            currVal = max(currVal - 1, values[index] - 1)
        
        return result

File: Dynamic Programming/Java/Best Sightseeing Pair.java
// Problem link: https://leetcode.com/problems/best-sightseeing-pair/description
// Time: O(n)
// Space: O(1)
class Solution {
    public int maxScoreSightseeingPair(int[] values) {
        int result = 0, currVal = values[0] - 1;

        for(int index = 1; index < values.length; index++) {
            result = Math.max(result, currVal + values[index]);
            currVal = Math.max(currVal - 1, values[index] - 1);
        }

        return result;
    }
}
